src/appleseed/foundation/meta/tests/test_imagefilewriter.cpp

```diff
@@ -0,0 +1,245 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2019-2020 Chin-Ying Li, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.foundation headers.
+#include "foundation/image/color.h"
+#include "foundation/image/colorspace.h"
+#include "foundation/image/genericimagefilereader.h"
+#include "foundation/image/genericimagefilewriter.h"
+#include "foundation/image/image.h"
+#include "foundation/image/imageattributes.h"
+#include "foundation/image/pixel.h"
+#include "foundation/math/rng/distribution.h"
+#include "foundation/math/rng/mersennetwister.h"
+#include "foundation/math/scalar.h"
+#include "foundation/utility/iostreamop.h"
+#include "foundation/utility/test.h"
+
+// Standard headers.
+#include <cassert>
+#include <cmath>
+#include <cstddef>
+#include <cstring>
+#include <memory>
+
+using namespace foundation;
+
+TEST_SUITE(Foundation_Image_ImageFileWriter)
+{
+    TEST_CASE(WriteBMP_CorrectlyWritesImagePixels)
+    {
+        const char* ImageFilePath = "unit tests/outputs/test_imagefilewriter.bmp";
+        
+        {
+            Image image(2, 2, 2, 2, 3, PixelFormatUInt16);
```

Comment:
You're using 16-bit integer channels for your images but that's not supported by BMP. That may be the reason why you're getting differences between what you write and what you read back.

Reply:
Using PixelFormatUInt16 works for BMP, and there's no discrepancy. But is this unexpected behaviour?

Reply:
BMP support 8 bits colors https://en.wikipedia.org/wiki/BMP_file_format

Take this for example : Write a gradient in 16 bit in a bmp, open it, the colors will be interpreted as 8bit and the gradient will not be the same

Reply:
The problem is the same with PNG and floating points